### PR TITLE
Jalali Calendar

### PR DESCRIPTION
**Pull request by RaianRaika**
New Feature-Add multi Calendar Logic to GnuCash 

in this commit, I add some concept and implementation of them to add a new feature that requested by the user that want to use a different calendar instance the gregorian calendar. 

the main idea in this implementation is:
1-I do not change the default Date (gregorian) in backend (saving and restoring) process, and just simply add a masked date ( convert from gregorian based user preferred Calendar selection) and show to the user. 
2-for user input, user can input the date directly , or use my DatePicker that added to this project, the DatePicker is based on Gnome Calendar. 
3- if anyone, want to add an other calendar to GnuCash, just need easyly add and impliment of my calendar-logic library for target Calendar.

### CODE REVIEW

**Comment by esahekmat** [comment]
great patch, lack of multi-calendar system in GnuCash is very annoying the users which works with calendar systems like Jalali, not Gregorian. also, I think this patch by the means of multi-calendar logic can be used to implementing other calendar systems (like Hebrew or Hijri)  in the future.

File: src/libqof/qof/gnc-datetime.hpp [hunk 1]
```diff
@@ -36,6 +37,13 @@ typedef struct
     int day; //1-31
 } ymd;
 
+typedef struct
```
Comment (jralls) [inline src/libqof/qof/gnc-datetime.hpp]: gnc-datetime.hpp is a C++-only file. ```typedef struct``` is C.

**Comment by jralls** [comment]
@RaianRaika,

The implementation in gnc-date and gnc-datetime is backwards. Implement everything in **C++11** using a class hierarchy to provide calendar classes. Renam GncDateTimeImpl to GncGregorian create a new GncJalali subclass of GncDateTime. Provide a factory function that picks which one to instantiate based on the preference setting. gnc-jalali.h/c should be changed to gnc-jalali.hpp/cpp and should implement the GncJalali calendar class, using whatever of Christian’s old code that  you like, but translated into C++.

 The whole thing should be completely invisible to the C functions in gnc-date.cpp; gdate-utils will have to be converted to C++ (it can move to libqof too) so that it can call GncDateTime to do what it needs, and that will fix the problem of OS-provided date functions. We can't use them because 32-bit and LLP64 systems have a 32-bit time_t that fails for dates after mid-2038. gnc-date's time64 exists purely to work around that, but in GncJalali you can use boost::datetime directly.

Then there's GncCalendar. Rather than copying the whole thing into GnuCash, you should subclass GtkCalendar and override only the functions that you need to.

Aside from the design problem, some general issues:

Please reformat your code to meet our [coding standards](http://wiki.gnucash.org/wiki/CodingStandard). You may find [artistic style](http://astyle.sourceforge.net/) helpful; we use ```astyle --indent=spaces=4 --brackets=break --suffix=none``` that cleans up most of the problems.

Block comments should be enclosed with ```/* … */```; ```/**…*/``` for Doxygen comments.

You declare ```static``` functions in header files. That’s a major no-no. ```static``` functions have scope only in the file they’re defined in.

File: src/app-utils/make-prefs-migration-script.xsl [hunk 2]
```diff
@@ -217,4 +233,9 @@
   <df:dateformat><df:name>locale</df:name><df:index>4</df:index></df:dateformat>
 </df:dateformats>
 
+<ct:calendartypes>
+    <ct:calendartype><ct:name>gregorian</ct:name><ct:index>0</ct:index></ct:calendartype>
+    <ct:calendartype><ct:name>jalali</ct:name><ct:index>1</ct:index></ct:calendartype>
+</ct:calendartypes>
+
 </xsl:stylesheet>
```
Comment (gjanssens) [inline src/app-utils/make-prefs-migration-script.xsl]: There is no need to add the calender type as a migratable pref. This feature is only used to migrate settings from gconf to dconf. Gconf was used until gnucash 2.4 so only preferences that already existed should be in here.

File: src/app-utils/migratable-prefs.xml.in [hunk 3]
```diff
@@ -386,6 +386,11 @@
       <gconfkey>24hour_clock</gconfkey>
     </pref>
     <pref>
+      <gschemaname>calendar-types</gschemaname>
+      <gschematype>calendartyp</gschematype>
+      <gconfkey>calendar_types</gconfkey>
+    </pref>
+    <pref>
```
Comment (gjanssens) [inline src/app-utils/migratable-prefs.xml.in]: Same here. This should not be added here.

**Comment by jralls** [comment]
Closing because the author hasn't responded to any of the comments in almost a year, so I guess it's abandoned.

**Pull request closed without merging**